ocaml/xenopsd/xc/service.ml

```diff
@@ -668,56 +668,90 @@ module Swtpm = struct
       debug "vTPM state for domid %d is empty: not restoring" domid
 
+  let check_state_needs_init task vtpm_uuid =
+    let dbg = Xenops_task.get_dbg task in
+    let contents =
+      Varstore_privileged_client.Client.vtpm_get_contents dbg vtpm_uuid
+    in
+    String.length contents = 0
+
   let start ~xs ~vtpm_uuid ~index task domid =
     debug "Preparing to start swtpm-wrapper to provide a vTPM (domid=%d)" domid ;
     let name = "swtpm" in
     let exec_path = !Resources.swtpm_wrapper in
     let pid_filename = pidfile_path domid in
     let vm_uuid = Xenops_helpers.uuid_of_domid ~xs domid |> Uuidx.to_string in
 
-    let chroot, _socket_path =
+    let chroot, socket_path =
       Xenops_sandbox.Swtpm_guard.start (Xenops_task.get_dbg task) ~vm_uuid
         ~domid ~paths:[]
     in
     let tpm_root =
       Xenops_sandbox.Chroot.(absolute_path_outside chroot Path.root)
     in
-    (* the uri here is relative to the chroot path, if chrooting is disabled then
-       swtpm-wrapper should modify the uri accordingly.
-       There are two modes in dictated by dir:// and file://.  The latter indicates
-       a linear file storage, and allows further permissions to be restricted.
-       xenopsd needs to be in charge of choosing the scheme according to the backend
+    (* the uri here is relative to the chroot path,
+       if chrooting is disabled then swtpm-wrapper should chdir accordingly.
+       There are three modes: dir://, file:// and unix+http://.
+       file:// indicates a linear file storage, and allows further permissions to be restricted.
+       xenopsd needs to be in charge of choosing the scheme according to the backend.
+
+       unix+http is the new backend, and dir:// is used during initial
+       manufacture.
     *)
-    let state_uri =
-      Filename.concat "dir://"
-      @@ Xenops_sandbox.Chroot.chroot_path_inside state_path
+    let args needs_init =
+      let state_uri =
+        if needs_init then
+          "dir://." (* exactly 2 //, we can't create this with Uri *)
+        else
+          (* drop leading / so that the URI works both inside and outside a
+             chroot *)
+          let host = String.sub socket_path 1 (String.length socket_path - 1) in
+          Uri.make ~scheme:"unix+http" ~host () |> Uri.to_string
+      in
+      Fe_argv.Add.many
+        [string_of_int domid; tpm_root; state_uri; string_of_bool needs_init]
+      |> Fe_argv.run
+      |> snd
+      |> Fe_argv.argv
     in
-    let args = Fe_argv.Add.many [string_of_int domid; tpm_root; state_uri] in
-    let args = Fe_argv.run args |> snd |> Fe_argv.argv in
+    (* fetch it here instead of swtpm-wrapper: better error reporting,
+       swtpm-wrapper runs as a service and getting the exact error back is
+       difficult. *)
+    let needs_init = check_state_needs_init task vtpm_uuid in
     let timeout_seconds = !Xenopsd.swtpm_ready_timeout in
+    if needs_init then (
+      debug "vTPM is empty, needs to be created" ;
```

Comment:
Should this include a path?

Reply:
no path anymore, but I will include the vtpm UUID.